docs/tutorial/source-plugin-tutorial.md

```diff
@@ -0,0 +1,977 @@
+
+Start by installing `gatsby-source-filesystem` in the `source-plugin` project:
+
+```shell:title=source-plugin
+npm install gatsby-source-filesystem
+```
+
+Now in your plugin's `gatsby-node.js` file, you can implement a new API, called `onCreateNode`, that gets called every time a node is created. You can check if the node created was one of your `Post` nodes, and if it was, create a file from the URL on the `imgUrl` field.
+
+Import the `createRemoteFileNode` helper from `gatsby-source-filesystem`, which will download a file from a remote location and create a `File` node for you.
+
+```javascript:title=source-plugin/gatsby-node.js
+const { ApolloClient } = require("apollo-client")
+const { InMemoryCache } = require("apollo-cache-inmemory")
+const { split } = require("apollo-link")
+const { HttpLink } = require("apollo-link-http")
+const { WebSocketLink } = require("apollo-link-ws")
+const { getMainDefinition } = require("apollo-utilities")
+const fetch = require("node-fetch")
+const gql = require("graphql-tag")
+const WebSocket = require("ws")
+// highlight-start
+const { createRemoteFileNode } = require(`gatsby-source-filesystem`)
+// highlight-end
+```
+
+Then export a new function `onCreateNode`, and call `createRemoteFileNode` in it whenever a node of of type `Post` is created:
+
+```javascript:title=source-plugin/gatsby-node.js
+// called each time a node is created
+exports.onCreateNode = async ({
+  node, // the node that was just created
+  actions: { createNode },
+  createNodeId,
+  getCache,
+}) => {
+  if (node.internal.type === POST_NODE_TYPE) {
+    const fileNode = await createRemoteFileNode({
+      // the url of the remote image to generate a node for
+      url: node.imgUrl,
+      parentNodeId: node.id,
+      createNode,
+      createNodeId,
+      getCache,
+    })
+
+    if (fileNode) {
+      node.remoteImage___NODE = fileNode.id
+    }
+  }
+}
+```
+
+This code is called every time a node is created, e.g. when `createNode` is invoked. Each time it is called in the `sourceNodes` step, the condition will check if the node was a `Post` node. Since those are the only nodes with an image associated with them, that is the only time images need to be optimized. Then a remote node is created, if it's successful, the `fileNode` is returned. The next few lines are important:
+
+```javascript:title=source-plugin/gatsby-node.js
+if (fileNode) {
+  // save the ID of the fileNode on the Post node
+  node.remoteImage___NODE = fileNode.id
+}
+```
+
+By assigning a field called `remoteImage___NODE` to the ID of the `File` node that was created, Gatsby will be able to [infer](/docs/glossary#inference) a connection between this field and the file node. This will allow fields on the file to be queried from the post node.
+
+```graphql
+# leaving off the ___NODE field will give you:
+query {
+  allPost {
+    nodes {
+      id
+      remoteImage # returns an ID like "ecd83d94-7111-5386-bd3f-0066248b6fa9"
```

Comment:
if i run `gatsby develop` before adding the `onCreateNode` it not get called, because the `Post` nodes already created and cached, so the `onCreateNode` is never called with `type=Post`

i had to do a gatsby clean to see the `remoteImage`...